adoc/chapters/programming_interface.adoc

```diff
@@ -21277,142 +21277,273 @@ requirements for both host and device.
 
 SYCL provides a number of function objects in the [code]#sycl# namespace on host
 and device.
-All function objects obey {cpp} conversion and promotion rules.
-Each function object is additionally specialized for [code]#void# as a
-_transparent_ function object that deduces its parameter types and return type.
 
-[source,,linenums]
+==== [code]#plus# class template
+
+[source,role=synopsis]
 ----
-include::{header_dir}/functional.h[lines=4..-1]
+include::{header_dir}/functional/plus.h[lines=4..-1]
 ----
 
-[[table.function.objects.plus]]
-.Member functions for the [code]#plus# function object
-[width="100%",options="header",separator="@",cols="65%,35%"]
-|====
-@ Member function @ Description
-a@
-[source]
+'''
+
+.[apititle]#call operator#
+[source,role=synopsis,id=api:plus-call-operator]
 ----
-T operator()(const T& x, const T& y) const
+T operator()(const T& x, const T& y) const;
 ----
-   a@ Returns the sum of its arguments, equivalent to [code]#pass:[x + y]#.
 
-|====
+_Returns_: [code]#pass:[x + y]#.
 
-[[table.function.objects.multiplies]]
-.Member functions for the [code]#multiplies# function object
-[width="100%",options="header",separator="@",cols="65%,35%"]
-|====
-@ Member function @ Description
-a@
-[source]
+'''
+
+.[apititle]#transparent call operator#
+[source,role=synopsis,id=api:plus-transparent-call-operator]
 ----
-T operator()(const T& x, const T& y) const
+template <class T, class U> constexpr auto operator()(T&& t, U&& u) const
+  -> decltype(std::forward<T>(t) + std::forward<U>(u));
```

Comment:
Is there a motivation to have the `std::forward` in the non-evaluated context?
Actually a maniac programmer could define `+` with an l-reference to do division and a `+` with an r-reference to do a multiplication?

Reply:
I don't know the motivation, but this is the way that the return type is declared for the equivalent function objects in in ISO C+ (e.g., https://eel.is/c++draft/function.objects#arithmetic.operations.plus-2).

I think you're right, though -- if `std::forward<T>(t) + std::forward<U>(u)` is the statement that the operator will run, it's safer to include the `std::forward` in the `decltype` just in case somebody does something really weird.

Reply:
Yes, I should remember that if it is weird in the C++ standard it is because some people do weird things in real life at the first place!